rr_viz/src/rr_viz_tabs.py

```diff
@@ -80,46 +84,86 @@ def __init__(self, parent):
+        # Set up topic subscribers
+        self.battery_level_sub_name = rospy.get_param("/battery/level", "/vesc_driver/battery")
+        self.battery_level_sub = rospy.Subscriber(self.battery_level_sub_name, BatteryState, self.battery_level_update)
 
+        # Set up core connection variables
+        self.core_clock_sub_name = rospy.get_param("/core_clock_sub","/core_clock_topic")
+        self.core_clock_sub = rospy.Subscriber(self.core_clock_sub_name, Time, self.core_clock_cb)
+        self.last_received_time = Time()
+
+        self.read_time = Time()
+        self.no_time_change = 0
+        self.latch = 0
+
+        # Core timer
+        core_clock_pub_freq = rospy.get_param("~publish_frequency", 10)
+        freq = float(core_clock_pub_freq) - (float(core_clock_pub_freq) / 100 * 10)
+        self.core_clock_timer_dur = rospy.Duration(1.0 / freq)
+        self.core_clock_timer = rospy.Timer(self.core_clock_timer_dur, self.core_clock_timer_cb, oneshot = True)
+        self.core_clock_timer_trig = False
+        
         # Set up timer 
-        self.timer_period = 1500
+        self.timer_period = 1000
         self.timer = QTimer(self)
         self.timer.timeout.connect(self.ros_is_up)
         self.timer.start(self.timer_period)
 
+        self.low_battery_popup.connect(self.bat_message_popup)
+
+    def core_clock_timer_cb(self, event):
+        self.core_clock_timer_trig = True
+        self.connection_status.setText("Lost")
+        self.connection_status.setStyleSheet("color: red")
+
+    def core_clock_cb(self, msg):
+        if self.core_clock_timer_trig:
+            self.core_clock_timer_trig = False
+        else:
+            self.core_clock_timer.shutdown()
+            self.core_clock_timer = rospy.Timer(self.core_clock_timer_dur, self.core_clock_timer_cb, oneshot = True)
+        
+        self.read_time = msg
+       
     def ros_is_up(self):
-        process = subprocess.Popen(self.subprocess_command, shell=True, stdout=subprocess.PIPE)
-        stdout = process.communicate()[0]
+        if self.last_received_time.data.secs == 0:
+            self.last_received_time = self.read_time
+            return
 
-        if not self.fp_status in stdout and not self.ros_loss_triggered:
-            self.connection_status.setText("Connection lost")
-            self.connection_status.setStyleSheet("color: red")
-            self.ros_loss_triggered = True
-        elif self.fp_status in stdout and self.ros_loss_triggered:
-            self.connection_status.setText("Connected")
+        if self.read_time.data.secs - self.last_received_time.data.secs == 0:
+            self.no_time_change += 1
+        else:
+            self.no_time_change = 0
+
+        if(self.no_time_change <= 3):
+            self.connection_status.setText("Good")
             self.connection_status.setStyleSheet("color: green")
-            self.ros_loss_triggered = False
+        elif self.no_time_change > 3 and self.no_time_change <= 5: 
```

Comment:
In here you can drop the first conditional as it's redundant:

```
        elif self.no_time_change <= 5: 
```

should have the same result